lib/conrad/emitters/kinesis.rb

```diff
@@ -0,0 +1,79 @@
+require 'conrad/errors'
+
+module Conrad
+  # A module containing all of conrad's built in event emitters for outputting
+  # events
+  module Emitters
+    # Basic emitter for sending events to AWS's kinesis event stream. The emitter will
+    # attempt to use values configured in the running environment according to
+    # the AWS SDK documentation (such as from ~/.aws/credentials).
+    class Kinesis
+      # Error for responding with issues around kinesis credential creation
+      class InvalidAwsCredentials < ::Conrad::Error
+        # :nodoc:
+        def to_s
+          'Must provide secret_access_key and access_key_id OR rely ' \
+          'on configured values in the running environment.'
+        end
+      end
+
+      # @return [String] the configured kinesis stream name
+      attr_reader :stream_name
+
+      # @return [String, nil] the configured aws region
+      attr_reader :region
+
+      # @return [Aws::Kinesis::Client] the created client
+      attr_reader :client
+
+      # @param stream_name [String] the name of the kinesis stream to send messages to
+      # @param region [String] region the stream lives in
+      # @param access_key_id [String] AWS Acesss Key ID
+      # @param secret_access_key [String] AWS Secret Access Key
+      #
+      # @raise [InvalidAwsCredentials] if access_key_id or secret_access_key are
+      #   not provided AND the running environment does not have valid AWS
+      #   credentials
+      # @raise [Aws::Errors::MissingRegionError] if region is not provided and
+      #   also not set via an allowed AWS environment variable
+      def initialize(stream_name:, region: nil, access_key_id: nil, secret_access_key: nil)
+        @stream_name = stream_name
+
+        create_client(region: region, access_key_id: access_key_id, secret_access_key: secret_access_key)
+      end
```

Comment:
this looks awfully similar to the sqs emitter. should we combine some of the shared functionality into a base class, concern, or a dependency we can inject?

Reply:
create_client definitely does, but I'm having a hard time thinking of a good way to pull it out since the meat of it is making an Aws::SOMETHING_HERE::Client and those all take different params. I think the SQS one wants a stream name too.

It feels like this is the smallest unique unit I can break it out to. What did you have in mind?

Reply:
unless i'm missing something it looks like shape is of `create_client` is identical. can you do something like
```ruby
# in the baseclass/concern
        #...
        create_client(region: region, access_key_id: access_key_id, secret_access_key: secret_access_key)
      end

      def create_client(region:, access_key_id:, secret_access_key:)
        if secret_access_key.nil? || access_key_id.nil?
          validate_implicit_credentials

          @client = self.class.client_class.new({ region: region }.compact)
        else
          @client = self.class.client_class.new({
            region: region,
            access_key_id: access_key_id,
            secret_access_key: secret_access_key
          }.compact)
        end
      end


# in Emitters::SQS
     def self.client_class
        Aws::SQS::Client
     end

# in Emitters::Kenesis 
     def self.client_class
        Aws::Kenesis::Client
     end

```
or did i miss something?

Reply:
Oh, sorry, I thought the call to initialize the sqs client took the queue_url, but the kinesis client waited until the event call to pass the stream name, which would mean they didn't have the same interface.

I'm a little wary of trying to pull the duplication out now cause we only got the two files and don't really know if there's a pattern with the other aws clients that'll be useful, but I think that's kind of a wash now and it wouldn't be hard to come back to what I got here if that's a problem, so I'm happy to do it. I'll push changes in a sec.